pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,129 @@
+// Currently the config is only generated for sidecar node on inbound HTTP/TCP listener. The generation
+// is controlled by ClusterRbacConfig (a singleton custom resource with cluster scope). User could disable
+// this plugin by either deleting the ClusterRbacConfig or set the ClusterRbacConfig.mode to OFF.
+// Note: ClusterRbacConfig is not created with default istio installation which means this plugin doesn't
+// generate any RBAC config by default.
+//
+// Changes from rbac_v2.go compared to rbac.go:
+// * Deprecate ServiceRoleBinding. Only support two CRDs: ServiceRole and AuthorizationPolicy.
+// * Allow multiple bindings and roles in one CRD, i.e. Authorization.
+// * Support workload selector.
+package authz
+
+import (
+	"fmt"
+
+	http_config "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/rbac/v2"
+	policyproto "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// convertRbacRulesToFilterConfigV2 is the successor of convertRbacRulesToFilterConfig, which supports
+// converting AuthorizationPolicy.
+func convertRbacRulesToFilterConfigV2(service *serviceMetadata, option rbacOption) *http_config.RBAC {
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	if option.globalPermissiveMode {
+		// Permissive Mode will be ignored.
+		rbacLog.Debugf("Permissive mode is not supported for service %s.", service.name)
+		// TODO(pitlv2109): Handle permissive mode in the future.
+		return &http_config.RBAC{Rules: rbac}
+	}
+
+	namespace := service.attributes[attrDestNamespace]
+	if option.authzPolicies == nil || option.authzPolicies.NamespaceToAuthorizationConfigV2 == nil {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	if _, present := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace]; !present {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	// Get all AuthorizationPolicy Istio config from this namespace.
+	allAuthzPolicies := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace].AuthzPolicies
+	for _, authzPolicy := range allAuthzPolicies {
+		// If a WorkloadSelector is used in the AuthorizationPolicy config and does not match this service, skip this
+		// AuthorizationPolicy.
+		serviceLabelsCollection := model.LabelsCollection{service.labels}
+		if authzPolicy.Policy.WorkloadSelector != nil &&
+			!(serviceLabelsCollection.IsSupersetOf(authzPolicy.Policy.WorkloadSelector.Labels)) {
+			continue
+		}
+		for i, binding := range authzPolicy.Policy.Allow {
+			// Check if we have converted this ServiceRole before.
+			permissions := make([]*policyproto.Permission, 0)
+			// Convert role.
+			role := option.authzPolicies.RoleForNameAndNamespace(namespace, binding.RoleRef.Name)
+			for _, rule := range role.Rules {
+				// Check to make sure the services field does not exist in the ServiceRole.
+				if len(rule.Services) > 0 {
```

Comment:
We may just ignore the services field considering we're already in the v2 path, this could the work in galley easier, then it doesn't need to modify ServiceRole at all.

Reply:
You mean just ignore the `services` field and do not skip a rule if it has the `services` field in it?

Reply:
yes

Reply:
Done.